File: crates/cli/src/serve.rs
//! `fat-horses serve`: the HTTP API on localhost for UI development (T4.5).
//! Picks run in-process in the background; the store is the CLI's JSON file.

use std::collections::HashMap;
use std::net::SocketAddr;
use std::path::PathBuf;
use std::sync::Arc;
use std::time::Duration;

use anyhow::Result;
use app::api::{Api, Method, Request, WorkflowStarter};
use app::workflow::{Config, Deps, SystemClock, run_pick};
use axum::Router;
use axum::body::Bytes;
use axum::extract::{Query, State};
use axum::http::{HeaderMap, StatusCode, Uri};
use axum::response::{IntoResponse, Json};
use axum::routing::any;
use domain::classify::FakeClassifier;
use domain::countries::CountriesFile;
use domain::store::PickStore;
use places::nominatim::{self, Nominatim};
use places::overpass::{self, Overpass};
use race::{Identity, TabNz};
use rand::SeedableRng;
use rand::rngs::StdRng;
use store::FileStore;

type Store = Arc<FileStore>;
type LocalApi = Api<Nominatim, Store, LocalStarter>;

/// Runs each pick as a background task in this process.
pub struct LocalStarter {
    store: Store,
    http: reqwest::Client,
}

impl WorkflowStarter for LocalStarter {
    async fn start(&self, pick_id: &str) -> Result<(), String> {
        let session = self
            .store
            .get_pick(pick_id)
            .await
            .map_err(|e| e.to_string())?
            .ok_or("pick not stored")?;
        let races = TabNz::new(race::tab_nz::DEFAULT_BASE_URL, Identity::from_env())
            .map_err(|e| e.to_string())?;
        let deps = Deps {
            races,
            places: Overpass::new(
                self.http.clone(),
                overpass::DEFAULT_ENDPOINTS
                    .iter()
                    .map(|s| s.to_string())
                    .collect(),
            ),
            classifier: FakeClassifier::new(),
            store: self.store.clone(),
            countries: CountriesFile::bundled(),
            config: Config::default(),
        };
        let id = pick_id.to_string();
        tokio::spawn(async move {
            let mut rng = StdRng::from_rng(&mut rand::rng());
            match run_pick(&deps, session, &SystemClock, &mut rng, |_| {}).await {
                Ok(s) => tracing::info!(pick_id = %id, status = ?s.status, "pick finished"),
                Err(e) => tracing::error!(pick_id = %id, error = %e, "pick failed"),
            }
        });
        Ok(())
    }
}

async fn handle(
    State(api): State<Arc<LocalApi>>,
    method: axum::http::Method,
    uri: Uri,
    Query(query): Query<HashMap<String, String>>,
    headers: HeaderMap,
    body: Bytes,
) -> impl IntoResponse {
    let method = match method {
        axum::http::Method::GET => Method::Get,
        axum::http::Method::POST => Method::Post,
        _ => Method::Other,
    };
    // `uri.path()` keeps percent-encoding; the handler decodes it.
    let path = uri
        .path()
        .strip_prefix("/api")
        .unwrap_or(uri.path())
        .to_string();
    let req = Request {
        method,
        path,
        query,
        api_key: headers
            .get("x-api-key")
            .and_then(|v| v.to_str().ok())
            .map(str::to_owned),
        body: (!body.is_empty()).then(|| String::from_utf8_lossy(&body).into_owned()),
    };
    let resp = api.handle(req, chrono::Utc::now()).await;
    (
        StatusCode::from_u16(resp.status).unwrap_or(StatusCode::INTERNAL_SERVER_ERROR),
        Json(resp.body),
    )
}

pub async fn serve(
    store: FileStore,
    port: u16,
    api_key: String,
    web_dir: Option<PathBuf>,
) -> Result<()> {
    let http = places::http_client(Duration::from_secs(60))?;
    let store = Arc::new(store);
    let api = Arc::new(Api {
        geocoder: Nominatim::new(http.clone(), nominatim::DEFAULT_BASE_URL),
        store: store.clone(),
        starter: LocalStarter { store, http },
        countries: CountriesFile::bundled(),
        api_key,
    });
    let mut app = Router::new()
        .route("/api/{*rest}", any(handle))
        .with_state(api);
    if let Some(dir) = web_dir {
        let index = dir.join("index.html");
        app = app.fallback_service(
            tower_http::services::ServeDir::new(dir)
                .fallback(tower_http::services::ServeFile::new(index)),
        );
    }
    let addr = SocketAddr::from(([127, 0, 0, 1], port));
    println!("Serving the API on http://{addr}/api");
    let listener = tokio::net::TcpListener::bind(addr).await?;
    axum::serve(listener, app).await?;
    Ok(())
}

File: server/src/cli/serve.ts
// `fat-horses serve`: the HTTP API (and optionally the built web UI) on localhost
// for development. Picks run in-process; the store is the CLI's JSON file.

import { createReadStream, existsSync, statSync } from "node:fs";
import { createServer, type IncomingMessage, type ServerResponse } from "node:http";
import { extname, join, normalize } from "node:path";
import { nominatimFromEnv, Overpass } from "../adapters/osm";
import { identityFromEnv, TabNz } from "../adapters/tabNz";
import { Api, type WorkflowStarter } from "../app/api";
import { defaultConfig, runPick, systemClock, type Deps } from "../app/workflow";
import { FakeClassifier } from "../domain/classify";
import { bundledCountries } from "../domain/countries";
import { systemRng } from "../domain/rng";
import type { PickSession } from "../domain/session";
import type { StateStore } from "../store/state";
import { log } from "../log";

/** Runs each pick as a background task in this process. */
export class LocalStarter implements WorkflowStarter {
  private readonly running = new Map<string, AbortController>();

  constructor(private readonly deps: Deps) {}

  async start(pickId: string): Promise<void> {
    const session = await this.deps.store.getPick(pickId);
    if (!session) throw new Error("pick not stored");
    this.run(session);
  }

  /**
   * Carry on with picks a previous server left unfinished (F13): a pick runs
   * inside the server, so stopping the server stops it mid-way.
   */
  async resume(store: StateStore): Promise<number> {
    const picks = await store.unfinishedPicks();
    for (const s of picks) {
      log.info("resuming pick", { pick_id: s.pick_id, status: s.status });
      this.run(s);
    }
    return picks.length;
  }

  private run(session: PickSession): void {
    const pickId = session.pick_id;
    if (this.running.has(pickId)) return;
    const abort = new AbortController();
    this.running.set(pickId, abort);
    void runPick(this.deps, session, systemClock, systemRng, () => {}, abort.signal)
      .then((s) => log.info("pick finished", { pick_id: pickId, status: s.status }))
      .catch((e: unknown) => log.error("pick failed", { pick_id: pickId, error: String(e) }))
      .finally(() => this.running.delete(pickId));
  }

  async cancel(pickId: string): Promise<void> {
    this.running.get(pickId)?.abort();
  }
}

const TYPES: Record<string, string> = {
  ".html": "text/html; charset=utf-8",
  ".js": "text/javascript",
  ".css": "text/css",
  ".json": "application/json",
  ".svg": "image/svg+xml",
  ".png": "image/png",
  ".woff2": "font/woff2",
};

function readBody(req: IncomingMessage): Promise<string> {
  return new Promise((resolve, reject) => {
    const chunks: Buffer[] = [];
    req.on("data", (c: Buffer) => chunks.push(c));
    req.on("end", () => resolve(Buffer.concat(chunks).toString("utf8")));
    req.on("error", reject);
  });
}

function serveStatic(webDir: string, path: string, res: ServerResponse): void {
  const safe = normalize(decodeURIComponent(path)).replace(/^(\.\.[/\\])+/, "");
  let file = join(webDir, safe);
  if (!file.startsWith(webDir) || !existsSync(file) || statSync(file).isDirectory()) {
    file = join(webDir, "index.html"); // SPA fallback
  }
  res.writeHead(200, { "content-type": TYPES[extname(file)] ?? "application/octet-stream" });
  createReadStream(file).pipe(res);
}

export function serve(store: StateStore, port: number, apiKey: string, webDir?: string): void {
  const deps: Deps = {
    races: new TabNz(identityFromEnv()),
    places: new Overpass(),
    classifier: new FakeClassifier(),
    store,
    countries: bundledCountries(),
    config: defaultConfig(),
  };
  const starter = new LocalStarter(deps);
  void starter.resume(store).then((n) => {
    if (n > 0) console.log(`Resumed ${n} unfinished pick${n === 1 ? "" : "s"}`);
  });
  const api = new Api({
    geocoder: nominatimFromEnv(),
    store,
    starter,
    countries: deps.countries,
    apiKey,
  });
  const server = createServer((req, res) => {
    const url = new URL(req.url ?? "/", "http://localhost");
    if (url.pathname.startsWith("/api/")) {
      void (async () => {
        try {
          const header = req.headers["x-api-key"];
          const r = await api.handle(
            {
              method: req.method ?? "GET",
              // Keep percent-encoding: take the raw path from req.url, not url.pathname.
              path: ((req.url ?? "").split("?")[0] ?? "").slice(4),
              query: Object.fromEntries(url.searchParams),
              apiKey: Array.isArray(header) ? header[0] : header,
              body: (await readBody(req)) || undefined,
            },
            new Date().toISOString(),
          );
          res.writeHead(r.status, { "content-type": "application/json" });
          res.end(JSON.stringify(r.body));
        } catch (e) {
          log.error("request failed", { error: String(e) });
          res.writeHead(500, { "content-type": "application/json" });
          res.end(JSON.stringify({ error: "internal", message: "unexpected error" }));
        }
      })();
    } else if (webDir) {
      serveStatic(webDir, url.pathname, res);
    } else {
      res.writeHead(404).end();
    }
  });
  server.listen(port, "127.0.0.1", () => console.log(`Serving the API on http://127.0.0.1:${port}/api`));
}
